### PR TITLE
Support external source Redis

### PR DESCRIPTION
**Pull request by comunodi**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category:
- New Feature

Short description (up to few sentences):
Support Redis storage as external dictionary via Poco driver

### CODE REVIEW

**Comment by proller** [comment]
Is this Still [WIP] ?

**Comment by alexey-milovidov** [comment]
Tests must be converted to integration tests, because dictionary tests were deleted.

**Comment by alexey-milovidov** [comment]
You have to tell your experience and performance testing results with Redis dictionaries.

**Comment by alesapin** [comment]
I have added integration test, but Redis dictionary doesn't seems to be ok. https://github.com/yandex/ClickHouse/commit/22fc9a96b30721869724a0ce630ded701f5d3ede.

File: dbms/src/Dictionaries/RedisBlockInputStream.cpp [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+            else
+                insertValue(*columns[idx], description.types[idx].first, value, name);
+        };
+
+        size_t num_rows = 0;
+        Poco::Redis::Command commandForValues("MGET");
+
+        while (num_rows < max_block_size)
+        {
+            if (cursor == keys.size())
```
Comment (alesapin) [inline dbms/src/Dictionaries/RedisBlockInputStream.cpp]: Increment of index after comparison  == segmentation fault.

File: dbms/src/Dictionaries/RedisBlockInputStream.cpp [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+            return {};
+
+        Poco::Redis::Array values = client->execute<Poco::Redis::Array>(commandForValues);
+        for (size_t i = 0; i < num_rows; ++i)
+        {
+            const Poco::Redis::RedisType::Ptr & value = *(values.begin() + i);
+            if (value.isNull())
+                insertDefaultValue(*columns[1], *description.sample_block.getByPosition(1).column);
+            else
+                insertValueByIdx(1, value);
```
Comment (alesapin) [inline dbms/src/Dictionaries/RedisBlockInputStream.cpp]: Why do we insert in first column of block only?

File: dbms/src/Dictionaries/RedisBlockInputStream.cpp [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+        }
+
+        return description.sample_block.cloneWithColumns(std::move(columns));
```
Comment (alesapin) [inline dbms/src/Dictionaries/RedisBlockInputStream.cpp]: Where is check, that columns consistent with sample block?

File: dbms/src/Dictionaries/RedisBlockInputStream.cpp [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+    }
+
+
+    Block RedisBlockInputStream::readImpl()
+    {
+        if (all_read)
+            return {};
+
+        const size_t size = 2;
+        assert(size == description.sample_block.columns());
```
Comment (alesapin) [inline dbms/src/Dictionaries/RedisBlockInputStream.cpp]: What does it mean? Why only two columns?

File: dbms/src/Dictionaries/RedisDictionarySource.cpp [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    {
+        auto createTableSource = [=](const DictionaryStructure & dict_struct,
+                                     const Poco::Util::AbstractConfiguration & config,
+                                     const std::string & config_prefix,
+                                     Block & sample_block,
+                                     const Context & /* context */) -> DictionarySourcePtr {
+#if USE_POCO_REDIS
+            return std::make_unique<RedisDictionarySource>(dict_struct, config, config_prefix + ".redis", sample_block);
+#else
+            (void)dict_struct;
```
Comment (alesapin) [inline dbms/src/Dictionaries/RedisDictionarySource.cpp]: Formating, style guide.

File: dbms/src/Dictionaries/RedisDictionarySource.cpp [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+
+    BlockInputStreamPtr RedisDictionarySource::loadIds(const std::vector<UInt64> & ids)
+    {
+        if (!dict_struct.id)
+            throw Exception{"'id' is required for selective loading", ErrorCodes::UNSUPPORTED_METHOD};
+
+        Poco::Redis::Array keys;
+
+        for (const UInt64 id : ids)
```
Comment (alesapin) [inline dbms/src/Dictionaries/RedisDictionarySource.cpp]: redundant const.

File: dbms/src/Dictionaries/RedisBlockInputStream.h [hunk 7]
```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+#include <Core/Block.h>
+#include <DataStreams/IBlockInputStream.h>
+#include "ExternalResultDescription.h"
```
Comment (alesapin) [inline dbms/src/Dictionaries/RedisBlockInputStream.h]: Doesn't compiles. This header in Common directory.

File: dbms/src/Dictionaries/RedisBlockInputStream.cpp [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+            commandForValues.addRedisType(key);
+        }
+
+        if (num_rows == 0)
+            return {};
+
+        Poco::Redis::Array values = client->execute<Poco::Redis::Array>(commandForValues);
+        for (size_t i = 0; i < num_rows; ++i)
+        {
+            const Poco::Redis::RedisType::Ptr & value = *(values.begin() + i);
```
Comment (alesapin) [inline dbms/src/Dictionaries/RedisBlockInputStream.cpp]: Unsafe. Why we expect values size to be equals to num_rows?

**Review by alesapin — CHANGES_REQUESTED** [review]

**Pull request merged**